### PR TITLE
Partial Movetables: allow moving a table one shard at a time

### PR DESCRIPTION
**Pull request by rohit-nayak-ps**
> **Warning!&nbsp;**
> This feature is expected to be used only for migrating large Vitess clusters across data centers and should not be enabled on an ongoing basis since it adds performance overhead during query serving. It also changes the paradigm hitherto expected in Vitess that a table is served, at any given time, from a single keyspace. The latter might lead to unstable behavior, with global routing, for example.

### TL;DR;

* MoveTables can now take a subset of shards
* ShardRoutingRules have been added, to route shards already moved from source to target
* ShardRoutingRules apply both for shard targeted and globally routed queries

## Description

This feature introduces the concept of partial keyspaces where some shards are served from a different keyspace. This is
useful for a specific but critical use-case where a large production Vitess setup (100s of shards) is being migrated to
a new data center or provider. Migrating the entire cluster in one go using `MoveTables` could cause an unacceptable
downtime due to the large number of primaries that need to be synced when writes are switched.

#### _partial_ MoveTables signalled by `--source_shards`

`vtctlclient MoveTables -- -source customer --tables 'customer,corder' --source_shards '-80' Create customer2.partial1`

#### VDiff works as-is

`vtctlclient VDiff customer2.partial1`

#### SwitchTraffic generates this shard routing rule
`vtctlclient MoveTables -- SwitchTraffic customer2.partial1`

`{"rules":[{"from_keyspace":"customer", "to_keyspace":"customer2", "shards":"-80"}]}`

#### Demo that shard routing now works

```
# stop workflow to stop reverse replication from running 
vtctlclient Workflow customer2.partial stop
# update customer2 database directly.
mysql -S $VTDATAROOT/vt_0000000500/mysql.sock -u vt_dba -e "update vt_customer2.customer set email = concat('routed.', email)"

# use shard targeting using customer (not customer2). 
mysql -e "use customer:-80; select customer_id, email from customer order by customer_id"

+-------------+---------------------------+
| customer_id | email                     |
+-------------+---------------------------+
|           1 | routed.alice@domain.com   |
|           2 | routed.bob@domain.com     |
|           3 | routed.charlie@domain.com |
|           5 | routed.eve@domain.com     |
+-------------+---------------------------+

```

# Summary of code changes

### Workflow Show changes
While we had the possibility of partial reads being switched earlier, now writes can also be partially switched in a workflow.

#### Topo 
Shard Routing Rules are a new concept introduced for this feature. It maps a (keyspace, shard) tuple to another keyspace. This is used to create a new cluster level map which maps a (keyspace, shard) to another keyspace. These are set by SwitchTraffic in a partial MoveTables and  used by vtgate while routing queries. vtctlclient commands `ApplyShardRoutingRules` and `GetShardRoutingRules`allow setting/getting of these rules.

#### vtgate Shard Targeting
The shard targeted query routing in vtgate's bypass mechanism `go/vt/vtgate/planbuilder/bypass.go`. We create a map from the SrvVSchema's shard routing rules object and check if a specified shard destination needs to be rerouted.

#### vtgate Global Routing
The global query routing using vtgate's ResolveDestinations() methods `go/vt/vtgate/vcursor_impl.go`. We go through all selected shard destinations and modify those that are mapped in the shard routing rules.

### 

### New column workflow_sub_type in _vt.vreplication
There is a new bool column `workflow_sub_type` added to `_vt.vreplication`, set for partial movetables. It is used for
visibility and for bypassing certain validations that expect a full keyspace.

#### vtgate --enable_shard_routing
Default: false. It is used when a cluster is setup with shard routing rules and tells vtgate to use these rules while routing queries.

#### MoveTables --source_shards
The only flag needed to tell MoveTables to perform a partial movetables is to pass it this flag, example `--source_shards -80`. This flag already exists and is used by Reshard.

## Notes

Both read and write traffic is switched at the same time when a shard routing is deemed complete (using SwitchTraffic). This is because we add a shard routing rule when this happens. Switching of read and write separately is done by updating the regular routing rules by targeting @replica and @rdonly.

#### e2e test TestPartialMoveTables
It creates a workflow which moves tables from one shard only to the target shard, ensures it completed correctly and switches traffic. It ensures that the shard routing rules and regular routing rules are setup correctly and that the vtgate queries both shard targeted and global queries are routed correctly.

#### Modified vtgate tests
A new set of tests have been added where the test cluster is setup as a partial cluster: one shard is in a different keyspace with the ShardRoutingRules setup. The main change is that the vtgate params need to specify the default keyspace as the `DbName` to avoid ambiguity.

Some tests have to be skipped for partial keyspaces because they intrinsically affect tablets in the base cluster which are now not all serving. 

TODOs:
[ ] Add explicit reasons for Skipped vtgate unit tests during partial keyspace

## Checklist
- [ ] Should this PR be backported?
- [X] Tests were added or are not required
- [ ] Documentation was added or is not required

### CODE REVIEW

File: test.go
Comment (mattlord) [inline test.go:179]: Need to use double dashes (single dash will be unsupported soon).

**Review by mattlord — COMMENTED** [review]
Looks great! Nice work! 

I only had some minor requests, comments, questions.

File: go/test/endtoend/vtgate/queries/insert/insert_test.go
Comment (harshit-gangal) [reply go/test/endtoend/vtgate/queries/insert/insert_test.go:62]: I think the skip message can be more clear as to why this particular test is skipped.
So, I would say that each skip message should be explain the reason for skipping that test.

File: go/test/endtoend/vtgate/errors_as_warnings/main_test.go
Comment (harshit-gangal) [inline go/test/endtoend/vtgate/errors_as_warnings/main_test.go:104]: we should skip only in case of partial_keyspace testing.

File: go/test/endtoend/vtgate/errors_as_warnings/main_test.go
Comment (harshit-gangal) [inline go/test/endtoend/vtgate/errors_as_warnings/main_test.go:137]: as we will be skipping the test for partial_keyspace, this change would not be needed.

File: go/test/endtoend/vtgate/prefixfanout/main_test.go
Comment (harshit-gangal) [reply go/test/endtoend/vtgate/prefixfanout/main_test.go:159]: all these should be removed in this file

File: go/test/endtoend/vtgate/queries/informationschema/informationschema_test.go
Comment (harshit-gangal) [reply go/test/endtoend/vtgate/queries/informationschema/informationschema_test.go:61]: adding reason for skipping will help

Comment (harshit-gangal) [reply go/test/endtoend/vtgate/queries/informationschema/informationschema_test.go:61]: also there are other places as well in this file

File: go/vt/vtgate/vtgate.go
Comment (harshit-gangal) [inline go/vt/vtgate/vtgate.go:309]: I think we should not set this like this.
instead in planbuilder we can call into vcusor to check if the flag is enabled or disabled.

File: go/vt/vtgate/vcursor_impl.go
Comment (harshit-gangal) [inline go/vt/vtgate/vcursor_impl.go:577]:
```suggestion
// is in the middle of being moved to another keyspace using MoveTables moving subset of shards at a time
```

File: go/vt/topotools/routing_rules_test.go
Comment (mattlord) [inline go/vt/topotools/routing_rules_test.go:80]: Should be using err and not srr:
```
require.NoError(t, err, "could not save shard routing rules to topo: %v", err)
```

File: go/vt/topotools/routing_rules_test.go
Comment (mattlord) [inline go/vt/topotools/routing_rules_test.go:83]:
```
require.NoError(t, err, "could not fetch shard routing rules from topo: %v", err)
```

File: go/vt/vtctl/workflow/stream_migrator.go
Comment (mattlord) [inline go/vt/vtctl/workflow/stream_migrator.go:577]:
```
int64(vrs.WorkflowType), int64(vrs.WorkflowSubType))
```

File: go/vt/wrangler/materializer.go
Comment (mattlord) [inline go/vt/wrangler/materializer.go:960]: If sourceShards2 is empty here... is that an error or does it just mean that it's not in fact partial?

File: go/vt/wrangler/materializer.go
Comment (mattlord) [inline go/vt/wrangler/materializer.go:976]: Same question here for targetShards2

File: go/vt/wrangler/materializer.go
Comment (mattlord) [inline go/vt/wrangler/materializer.go:980]: Ah, this answers my previous two questions :). IMO it would be better to error for each one as we return sooner (do less work) and we can provide a more specific error.

File: go/vt/wrangler/traffic_switcher.go
Comment (mattlord) [inline go/vt/wrangler/traffic_switcher.go:1456]: The log messages should be past tense, `Deleted`. Same for the one below: `Added`

File: go/vt/wrangler/traffic_switcher.go
Comment (mattlord) [inline go/vt/wrangler/traffic_switcher.go:1475]: Another one that should be past tense

**Review by mattlord — COMMENTED** [review]
Looks good! I had some minor questions/comments/suggestions. We also have some unresolved comments from @harshit-gangal to resolve in one way or another. I can approve quickly once we go through those things and resolve them. Thanks! ❤️

File: go/test/endtoend/vtgate/misc_test.go
Comment (rohit-nayak-ps) [reply go/test/endtoend/vtgate/misc_test.go:399]: done

File: go/vt/topotools/routing_rules.go
Comment (rohit-nayak-ps) [reply go/vt/topotools/routing_rules.go:65]: The topo server is passed down by the traffic switcher which has already referred to it, so if it is nil it should have been already reported (or panicked!) before it comes here.

File: go/vt/vtgate/planbuilder/bypass.go
Comment (rohit-nayak-ps) [reply go/vt/vtgate/planbuilder/bypass.go:29]: The set  is done as part of vtgate.Init() and get is done once vtgate comes up. I had originally added a mutex but realized later that they can't be called concurrently..

File: go/vt/vtgate/vindexes/vschema.go
Comment (rohit-nayak-ps) [reply go/vt/vtgate/vindexes/vschema.go:66]: I couldn't understand where the looping that is referred to: maybe the code has changed since this comment.

File: go/vt/vtgate/vindexes/vschema.go
Comment (rohit-nayak-ps) [reply go/vt/vtgate/vindexes/vschema.go:533]: The len() is on a separate (parent) object that the nil check.

File: go/vt/wrangler/materializer.go
Comment (rohit-nayak-ps) [reply go/vt/wrangler/materializer.go:960]: fixed

File: go/vt/wrangler/materializer.go
Comment (rohit-nayak-ps) [reply go/vt/wrangler/materializer.go:980]: fixed

File: go/vt/wrangler/resharder.go
Comment (rohit-nayak-ps) [reply go/vt/wrangler/resharder.go:340]: Yes, we need to query the vrep table. A lot of tests started failing when I did that initially, We don't really use that field other than for reporting in Workflow Show, so I will do this in a separate PR once this one is merged.

File: go/vt/wrangler/traffic_switcher.go
Comment (rohit-nayak-ps) [reply go/vt/wrangler/traffic_switcher.go:934]: They are two different Enums. I only changed the case in the WorkflowType in this PR.

File: .github/workflows/cluster_endtoend_vreplication_partialmovetables.yml
Comment (rohit-nayak-ps) [reply .github/workflows/cluster_endtoend_vreplication_partialmovetables.yml]: Changed it to use mysql80

File: .github/workflows/cluster_endtoend_vtgate_transaction_partial_keyspace.yml
Comment (rohit-nayak-ps) [reply .github/workflows/cluster_endtoend_vtgate_transaction_partial_keyspace.yml]: Done. I also refactored test.go to allow specifying multiple config.*json. The config.json file was becoming pretty long and this will allow specifying tests in a modular way.

File: go/cmd/vtctldclient/command/shard_routing_rules.go
Comment (rohit-nayak-ps) [reply go/cmd/vtctldclient/command/shard_routing_rules.go:95]: yes

File: test/config_partial_keyspace.json
Comment (mattlord) [inline test/config_partial_keyspace.json:1]: A lot of these have retries, do we expect them to be flaky?

**Review by mattlord — CHANGES_REQUESTED** [review]
Nice work on this! There were a couple of things we should change before merging (example related), but otherwise it's a few minor questions and suggestions that you can make the final call on. I'll review your feedback tomorrow and quickly approve. Thanks!

File: test/config_partial_keyspace.json
Comment (rohit-nayak-ps) [reply test/config_partial_keyspace.json:1]: I used the same settings as the original vtgate tests.

**Review by mattlord — APPROVED** [review]
❤️

**Pull request merged**
